pkg/api/norman/customization/gke/handler.go

```diff
@@ -0,0 +1,208 @@
+			logrus.Debugf("[gke-handler] error getting versions: %v", err)
+			handleErr(writer, errCode, err)
+			return
+		}
+		writer.Write(serialized)
+	case "gkeZones":
+		if serialized, errCode, err = listZones(req.Context(), capa); err != nil {
+			logrus.Debugf("[gke-handler] error getting zones: %v", err)
+			handleErr(writer, errCode, err)
+			return
+		}
+		writer.Write(serialized)
+	default:
+		handleErr(writer, httperror.NotFound.Status, fmt.Errorf("invalid endpoint "+resourceType))
+	}
+}
+
+func (h *handler) getCloudCredential(req *http.Request, cap *Capabilities, credID string) error {
+	ns, name := ref.Parse(credID)
+	if ns == "" || name == "" {
+		logrus.Debugf("[GKE] invalid cloud credential ID %s", credID)
+		return fmt.Errorf("invalid cloud credential ID %s", credID)
+	}
+
+	var accessCred client.CloudCredential //var to check access
+	if err := access.ByID(h.generateAPIContext(req), &schema.Version, client.CloudCredentialType, credID, &accessCred); err != nil {
+		if apiError, ok := err.(*httperror.APIError); ok {
+			if apiError.Code.Status == httperror.PermissionDenied.Status || apiError.Code.Status == httperror.NotFound.Status {
+				return fmt.Errorf("cloud credential not found")
+			}
+		}
+		return err
+	}
+
+	cc, err := h.secretsLister.Get(ns, name)
+	if err != nil {
+		logrus.Debugf("[GKE] error accessing cloud credential %s", credID)
+		return fmt.Errorf("error accessing cloud credential %s", credID)
+	}
+	cap.Credentials = string(cc.Data["googlecredentialConfig-authEncodedJson"])
+
+	cap.ProjectID = req.URL.Query().Get("projectId")
+	if cap.ProjectID == "" {
+		logrus.Debugf("[GKE] error getting projectId")
+		return fmt.Errorf("error getting projectId")
+	}
+
+	region := req.URL.Query().Get("region")
+	if region != "" {
+		cap.Region = region
+	}
+	zone := req.URL.Query().Get("zone")
+	if zone != "" {
+		cap.Zone = zone
+	}
+
+	return nil
+}
+
+func (h *handler) getCredentialsFromBody(writer http.ResponseWriter, req *http.Request, cap *Capabilities) error {
+	raw, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		writer.WriteHeader(http.StatusBadRequest)
+		return fmt.Errorf("cannot read request body: " + err.Error())
+	}
+
+	if err = json.Unmarshal(raw, &cap); err != nil {
+		writer.WriteHeader(http.StatusBadRequest)
+		return fmt.Errorf("cannot parse request body: " + err.Error())
+	}
+
+	if cap.Credentials == "" {
+		writer.WriteHeader(http.StatusBadRequest)
+		return fmt.Errorf("invalid credentials")
+	}
+	if cap.ProjectID == "" {
+		writer.WriteHeader(http.StatusBadRequest)
+		return fmt.Errorf("invalid projectId")
+	}
+
+	return nil
+}
+
+func (h *handler) generateAPIContext(req *http.Request) *types.APIContext {
+	return &types.APIContext{
+		Method:  req.Method,
+		Request: req,
+		Schemas: h.schemas,
+		Query:   map[string][]string{},
+	}
+}
+
+func handleErr(writer http.ResponseWriter, errorCode int, originalErr error) {
+	resp := errorResponse{originalErr.Error()}
+
+	asJSON, err := json.Marshal(resp)
```

Comment:
nit: format to `[]byte` directly and avoid the necessity of error checking and the `errorReponse` type.
```suggestion
	asJSON := []byte(fmt.Sprintf(`{"error":"%v"}`, originalErr))
```

Reply:
thank you, it looks great